### PR TITLE
Onboarding local storage

### PR DESCRIPTION
**Pull request by kimchouard**

### CODE REVIEW

File: app/src/reducers/index.js [hunk 1]
```diff
@@ -18,4 +19,5 @@ export default combineReducers({
   profile,
   welcome,
   walkthrough,
+  onboarding
```
Comment (kimchouard) [inline app/src/reducers/index.js]: I'd rename this to "isOnboarded" to be more explicit of what it is
```suggestion
  isOnboarded,
```

File: app/src/scripts/game/components/game.jsx [hunk 2]
```diff
@@ -21,6 +21,10 @@ class Game extends Component {
 
   componentDidMount() {
     this.props.stopWalkthrough();
+    console.log(this.props.getOnboarding());
```
Comment (kimchouard) [inline app/src/scripts/game/components/game.jsx]: This is not the right way to use your function and state.

--> getOnboarding() will fetch into chrome's google storage asynchronously, and then update your redux state (i.e. this.props...) once it the promise is completed.
--> getOnboarding() does NOT return any values. It's really important. It just updates the states. The way you get values should ALWAYS be via this.props..., not any functions.
--> You need to use `this.props.isOnboarded` to know whether the user is onboarded or not.
 
The value of `this.props.isOnboarded`  can be:
- `undefined`:  when the component is mounted, `this.props.isOnboarded` doesn't have any value. That means that the promise hasn't been returned yet.
- `false`: this means that the value has been returned from chrome storage, but is `false`, meaning that first walkthrough hasn't been done.
- `true`: that'll be the case most of the time. It means that the value has been returned from chrome storage and that the user has been onboarded.

```suggestion
    this.props.getOnboarding();
```

File: app/src/scripts/game/components/game.jsx [hunk 3]
```diff
@@ -21,6 +21,10 @@ class Game extends Component {
...
+    if (this.props.getOnboarding() === false) {
```
Comment (kimchouard) [inline app/src/scripts/game/components/game.jsx]: Remove all this if all together. Just run the function when you mount the component to make sure you're getting the value from storage.

Comment (kimchouard) [reply app/src/scripts/game/components/game.jsx]: Note: this actually needs to be in the rendering function, with the if a bit different.

File: app/src/scripts/game/components/walkthrough.jsx [hunk 4]
```diff
@@ -298,7 +303,7 @@ const mapStateToProps = (state) => {
 };
```
Comment (kimchouard) [inline app/src/scripts/game/components/walkthrough.jsx]: **You need to add `isOnboarded` in the `mapStateToProps` function, for it to be available in `this.props.isOnboarded`.**

File: app/src/scripts/game/components/game.jsx (hunk 3 above)
Comment (kimchouard) [reply app/src/scripts/game/components/game.jsx]: at the begining of the rendering function:

if ([onboarded value is returned & user not onboarded yet) {
  [start the onboarding]
}

**Comment by LtAstros** [comment]
Updates were made in the recent commit to fix the code to fit the comments made.

**Comment by LtAstros** [comment]
Another update was made to this brach. Also a discovery has been found of a bug that only starts the onboarding tutorial only after two refreshes.

**Comment by LtAstros** [comment]
Updated made as stated by the commit messages.

File: app/src/scripts/_utils.js [hunk 5]
```diff
@@ -81,7 +81,103 @@ export const addComplete = (quests, valleyName) => {
...
+    index: 3,
+    name: "Sharing",
+    image: null,
+    requirements: {
+      Placeholder: true
+    }
+  }
+];
+
+// export const getAge = (quests, agesData) => {
```
Comment (kimchouard) [inline app/src/scripts/_utils.js]: @LtAstros can we remove the comments for the final PR?

File: app/src/scripts/_utils.js [hunk 6]
```diff
@@ -81,7 +81,103 @@ export const addComplete = (quests, valleyName) => {
...
+//     if ( (newArray >= Object.values(ageData.requirements) && Object.values(ageData.requirements) !== -1) || newArray === Object.values(ageData.requirements)) {
+//       console.log("Passed Age: " + ageData.name)
+//     } else {
+//       console.log("Failed Age: " + ageData.name)
+//     }
+//     console.log("A whole new array " + newArray)
+//   });
+// };
+
+export const getAge = (quests, agesData) => {
```
Comment (kimchouard) [inline app/src/scripts/_utils.js]: You don't need to pass the agesData if it's already set in the file right?

Comment (kimchouard) [reply app/src/scripts/_utils.js]:
```suggestion
export const getAge = (quests) => {
```

File: app/src/scripts/game/components/header.jsx [hunk 7]
```diff
@@ -1,6 +1,6 @@
 import React, { Component } from 'react';
 import {connect} from 'react-redux';
-import { questUnlocked, addComplete } from '../../_utils';
+import { questUnlocked, addComplete, getAge, agesData } from '../../_utils';
```
Comment (kimchouard) [inline app/src/scripts/game/components/header.jsx]: Again, agesData would be a `private` variable, only used bt the utils function directly. So I'd not export that here.

```suggestion
import { questUnlocked, addComplete, getAge } from '../../_utils';
```

File: app/src/scripts/game/components/header.jsx [hunk 8]
```diff
@@ -18,8 +18,16 @@ class Header extends Component {
...
+
+  getRomanAge(ageData) {
+    if(ageData.index === 0) { return "I"; }
+    else if(ageData.index === 1) { return "II"; }
+    else if(ageData.index === 2) { return "III"; }
+    else if(ageData.index === 3) { return "IV"; }
+  }
   
   render() {
+    let ageData = getAge(this.props.quests,agesData);
```
Comment (kimchouard) [inline app/src/scripts/game/components/header.jsx]:
```suggestion
    let ageData = getAge(this.props.quests);
```

File: app/src/scripts/game/components/header.jsx [hunk 9]
```diff
@@ -18,8 +18,16 @@ class Header extends Component {
       this.props.startWalkthrough
     );
   }
+
+  getRomanAge(ageData) {
+    if(ageData.index === 0) { return "I"; }
```
Comment (kimchouard) [inline app/src/scripts/game/components/header.jsx]: Can we do a `case` statement here instead? :)

File: app/src/scripts/game/components/walkthrough.jsx [hunk 10]
```diff
@@ -19,7 +19,7 @@ class Walkthrough extends Component {
         this.props.stopWalkthrough();
         this.setState({stepIndex: 0});
         console.log("TOUR HAS STOPPED");
-
+        this.props.setOnboarding(true)
```
Comment (kimchouard) [inline app/src/scripts/game/components/walkthrough.jsx]: Why would you set this to true by default on the constructor?

File: app/src/scripts/_utils.js [hunk 11]
```diff
@@ -81,7 +81,103 @@ export const addComplete = (quests, valleyName) => {
...
+  //console.log(Object.keys(agesData[0].requirements))
+  for (let i = 0; i < agesData.length; i++) {
+    //console.log(ageData)
+    let newArray = Object.keys(agesData[i].requirements).map( (valleyName) => {
+      if (valleyName === "total") {
+        return addComplete(quests)
+      } else {
+        if (addComplete(quests, valleyName, null) === addComplete(quests, valleyName) && typeof(agesData[i].requirements[valleyName]) === "string"){
+          console.log()
```
Comment (kimchouard) [inline app/src/scripts/_utils.js]: can we remove this please?

File: app/src/scripts/_utils.js [hunk 12]
```diff
@@ -81,7 +81,103 @@ export const addComplete = (quests, valleyName) => {
...
+
+export const getAge = (quests, agesData) => {
+  //console.log(Object.keys(agesData[0].requirements))
+  for (let i = 0; i < agesData.length; i++) {
+    //console.log(ageData)
+    let newArray = Object.keys(agesData[i].requirements).map( (valleyName) => {
+      if (valleyName === "total") {
+        return addComplete(quests)
+      } else {
+        if (addComplete(quests, valleyName, null) === addComplete(quests, valleyName) && typeof(agesData[i].requirements[valleyName]) === "string"){
```
Comment (kimchouard) [inline app/src/scripts/_utils.js]: I'd check the typeof first for performance reasons:
```suggestion
        if (typeof(agesData[i].requirements[valleyName]) === "string" && addComplete(quests, valleyName, null) === addComplete(quests, valleyName)){
```

File: app/src/scripts/_utils.js [hunk 13]
```diff
@@ -81,7 +81,103 @@ export const addComplete = (quests, valleyName) => {
...
+          console.log()
+          return "all"
+        } else {
+          return addComplete(quests, valleyName)
+        }
+      }
+    });
+    console.log("A not so new array " + Object.values(agesData[i].requirements))
```
Comment (kimchouard) [inline app/src/scripts/_utils.js]:
```suggestion
    console.log("A not so new array ", Object.values(agesData[i].requirements))
```

**Pull request merged**